### ISSUE TITLE
[GPU] FlashAttention performance lags behind PyTorch

### ISSUE THREAD

**Issue by neel04** [jax-ml/jax#24934]
### Description

I'm benchmarking naive FlashAttention in `Jax` vs. the Pallas's version of [`FA3`](https://github.com/jax-ml/jax/blob/7b9914d711593dca8725d46aa1dadb2194284519/jax/experimental/pallas/ops/gpu/attention_mgpu.py#L14) vs. the new `dot_product_attention` interface with `cudnn` backend.

- Jax method' performance:  

![image](https://github.com/user-attachments/assets/beee91c1-f0fc-4987-8207-f547def5cf30)

- Torch's performance:  

![image](https://github.com/user-attachments/assets/18d981b5-f11f-4740-a3cf-b69bae3e5369)

Why the discrepancy? I'd have expected performance to touch 550-600 TFLOPS/s. I'm using a few XLA flags, as specified in the script below - but is there anything I'm missing? Or is this about the maximum `XLA` can deliver on H100s?

<details><summary>JAX script</summary>
<p>

```py
import os, sys

os.environ['XLA_PYTHON_CLIENT_MEM_FRACTION'] = '0.9'

os.environ['XLA_FLAGS'] = (
    '--xla_gpu_enable_triton_softmax_fusion=true '
    '--xla_gpu_triton_gemm_any=True '
    '--xla_gpu_enable_cudnn_fmha=true '
    '--xla_gpu_enable_triton_hopper=true '
)
import math
import time
from tabulate import tabulate
import jax
import jax.numpy as jnp
import numpy as np
from jax.tree_util import tree_map
from jax.experimental.pallas.ops.gpu.attention import mha as pallas_mha
from functools import partial

class Timer(object):
    def __init__(self, into=None):
        self.into = into

    def __enter__(self):
        self.start = time.time()

    def __exit__(self, type, value, traceback):
        if self.into is not None:
            self.into.append(time.time() - self.start)

    def elapsed(self):
        return time.time() - self.start

def cartesian(*lists):
    if lists:
        xs = lists[0]
        for x in xs:
            for rest in cartesian(*lists[1:]):
                yield (x,) + rest
    else:
        yield ()

def linear_flops_fwd(B, C1, C2):
    return 2 * B * C1 * C2
def linear_flops_bwd(B, C1, C2):
    return 4 * B * C1 * C2

def cross_attn_flops(B,T,TK,H,C):
    HC = H*C
    flops_fwd = (
        2 * B*H*T*TK*C + # S = Q@K
        3 * B*H*T*TK +  # P=softmax(S)
        2 * B*T*TK*H*C # O = P@V
    )
    return flops_fwd

def attn_flops(B,T,H,C):
    return cross_attn_flops(B,T,T,H,C)

dtype = jnp.float16
print(dtype)

def convert(xs):
    return tree_map(lambda x: x.astype(dtype), xs)

# ----

def ref_fwd(q,k,v):
    [n, l, h, d] = q.shape
    [n, lk, hk, d] = k.shape
    softmax_scale = 1/math.sqrt(d)
    S = jnp.einsum('nlhd,nLhd->nhlL',q,k)
    P = jax.nn.softmax(S*softmax_scale, axis=-1)
    o = jnp.einsum('nhlL,nLhd->nlhd',P,v)
    return o.astype(q.dtype)

def jax_dpa_fwd(q, k, v):
    # Reshape from [B, T, H, C] to [B, T, N, H] as expected by dot_product_attention
    B, T, H, C = q.shape
    scale = 1.0 / jnp.sqrt(C)
    
    # Call dot_product_attention without forcing CUDA
    output = jax.nn.dot_product_attention(
        query=q,
        key=k,
        value=v,
        implementation='cudnn'
    )
    
    return output

# ----

Bx = [32]
Tx = [1024]
Hx = [4, 8, 16, 32, 64]
Cx = [32, 64, 128]
sx = [8]

def bench_attn(mha):
    @jax.jit
    def bench(q, k, v, steps):
        def f(v, _):
            v = mha(q,k,v)
            return v, None
        out, _ = jax.lax.scan(f, v, steps, unroll=True)
        return out

    times = []
    table = {}

    for B,T,H,C,s in cartesian(Bx,Tx,Hx,Cx,sx):
        q = jax.random.normal(jax.random.PRNGKey(0), [B, T, H, C]).astype(dtype)
        k = jax.random.normal(jax.random.PRNGKey(1), [B, T, H, C]).astype(dtype)
        v = jax.random.normal(jax.random.PRNGKey(2), [B, T, H, C]).astype(dtype)
        steps = jnp.arange(s)

        out = []
        for _ in range(8):
            with Timer(out):
                bench(q, k, v, steps).block_until_ready()
        t = sum(out[1:])/len(out[1:])
        flop = attn_flops(B,T,H,C) * s
        tf = flop / t / 1e12
        print(f'flops ({B} {T} {H} {C} / {s}): {tf}T', out)
        table[(B,T,H,C,s)] = tf
        times.append((flop, t))

    # a @ x = b
    a = np.array([[flop, 1] for (flop,t) in times])
    b = np.array([[t] for (flop,t) in times])
    c = np.linalg.lstsq(a, b)[0]
    print(c)
    print(1/c[0]/1e12)
    print(c[1])
    return table

naive_flops = bench_attn(ref_fwd)
pallas_flops = bench_attn(partial(pallas_mha, segment_ids=None))
jax_dpa_flops = bench_attn(jax_dpa_fwd)

table = []
for B,T,H,C,s in cartesian(Bx,Tx,Hx,Cx,sx):
    table.append((B,T,H,C,naive_flops[(B,T,H,C,s)], pallas_flops[(B,T,H,C,s)], jax_dpa_flops[(B,T,H,C,s)]))

print(tabulate(table, headers=['B','T','H','C','TFlop/s (naive)','TFlop/s (pallas)','TFlop/s (jax_dpa)']))
```

</p>
</details> 

<details><summary>`PyTorch` Benchmark script</summary>
<p>

import torch
import time
from tabulate import tabulate
from typing import Optional
from itertools import product

class Timer:
    def __init__(self, into=None):
        self.into = into

    def __enter__(self):
        self.start = time.time()

    def __exit__(self, type, value, traceback):
        if self.into is not None:
            self.into.append(time.time() - self.start)

def attention_flops(B: int, T: int, H: int, C: int) -> int:
    """Calculate FLOPs for self-attention forward pass"""
    return (
        2 * B * H * T * T * C +  # Q @ K
        3 * B * H * T * T +      # softmax
        2 * B * T * T * H * C    # attn @ V
    )

class OptimizedMHA(torch.nn.Module):
    def __init__(self):
        super().__init__()
        
    def forward(self, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor) -> torch.Tensor:
        B, T, H, C = q.shape
        scale = 1.0 / (C ** 0.5)
        
        # Reshape for batch matrix multiplication
        q = q.transpose(1, 2)  # [B, H, T, C]
        k = k.transpose(1, 2)  # [B, H, T, C]
        v = v.transpose(1, 2)  # [B, H, T, C]
        
        # Compute attention scores
        attn = torch.matmul(q, k.transpose(-2, -1)) * scale  # [B, H, T, T]
        attn = torch.softmax(attn, dim=-1)
        
        # Apply attention to values
        out = torch.matmul(attn, v)  # [B, H, T, C]
        out = out.transpose(1, 2)  # [B, T, H, C]
        
        return out

def bench_attention():
    # Configuration
    device = torch.device("cuda")
    dtype = torch.float16
    
    # Test parameters
    Bx = [32]               # batch size
    Tx = [1024]            # sequence length
    Hx = [16, 32]    # number of heads
    Cx = [32, 64, 128]     # head dimension
    sx = [8]               # steps to run
    
    # Initialize model
    model = OptimizedMHA().to(device)
    compiled_model = torch.compile(
        model,
        mode="max-autotune",
        fullgraph=True,
        dynamic=False
    )
    
    results = []
    
    # Run benchmarks for each configuration
    for B, T, H, C, s in product(Bx, Tx, Hx, Cx, sx):
        # Create input tensors
        q = torch.randn(B, T, H, C, device=device, dtype=dtype)
        k = torch.randn(B, T, H, C, device=device, dtype=dtype)
        v = torch.randn(B, T, H, C, device=device, dtype=dtype)
        
        # Warmup
        for _ in range(10):
            with torch.no_grad():
                _ = compiled_model(q, k, v)
        torch.cuda.synchronize()
        
        # Benchmark
        times = []
        for _ in range(100):
            with torch.no_grad(), Timer(times):
                _ = compiled_model(q, k, v)
                torch.cuda.synchronize()
        
        # Calculate statistics
        times = times[10:]  # Discard first 10 runs
        avg_time = sum(times) / len(times)
        flops = attention_flops(B, T, H, C) * s
        tflops = flops / avg_time / 1e12
        
        print(f"Config (B={B}, T={T}, H={H}, C={C}): {tflops:.2f} TFlop/s")
        results.append((B, T, H, C, tflops))
    
    # Print results table
    headers = ['Batch', 'SeqLen', 'Heads', 'HeadDim', 'TFlop/s']
    print("\nResults:")
    print(tabulate(results, headers=headers, floatfmt='.2f'))

if __name__ == "__main__":
    bench_attention()

</p>
</details> 

### System info (python version, jaxlib version, accelerator, etc.)

```
>>> import jax; jax.print_environment_info()
jax:    0.4.35
jaxlib: 0.4.34
numpy:  2.1.3
python: 3.11.10 (main, Oct 16 2024, 04:38:48) [Clang 18.1.8 ]
device info: NVIDIA H100 PCIe-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='nifty-orthodox-whale', release='6.8.0-40-generic', version='#40~22.04.3-Ubuntu SMP PREEMPT_DYNAMIC Tue Jul 30 17:30:19 UTC 2', machine='x86_64')


$ nvidia-smi
Sun Nov 17 02:04:44 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.06             Driver Version: 535.183.06   CUDA Version: 12.4     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA H100 PCIe               On  | 00000000:00:07.0 Off |                    0 |
| N/A   31C    P0              54W / 350W |    467MiB / 81559MiB |      2%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
+---------------------------------------------------------------------------------------+
```

**Comment by neel04** [comment]
Update: I changed the `torch` script to use `FlopCounterMode`. Now the results are more realistic/accurate but JAX still lags behind despite explicitly being forced to use `CuDNN`.

cc @kaixih @sbodenstein @dfm

**Comment by Rick0827** [comment]
i think your torch script might not work as expected, since the inputs format of `torch.nn.functional.scaled_dot_product_attention` is [B, H, T, C] [https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html]
<img width="326" alt="63aaff67935ab7ebb1dc5feb4226787" src="https://github.com/user-attachments/assets/f1888de6-c1bd-40c1-9c22-f5c998c26d39">

**Comment by sbodenstein** [comment]
This is a complicated benchmarking setup, with many things potentially going wrong. Can you simplify this to just measuring milliseconds, and also have a correctness test (that PyTorch and JAX give the same output for the same input).

**Comment by neel04** [comment]
@Rick0827 Thank you for pointing that out.

@sbodenstein I have updated both scripts to now report times as well. However, I opted to skip correctness tests because reproducibility requires us to sacrifice performance which I'm afraid of touching

The variance however is very low between runs plus we can average over multiple steps (`sx`) so this should be a non-issue. 

On `A100`:

- **JAX**:

![image](https://github.com/user-attachments/assets/45755a25-aa30-4155-b9a1-4822f5f736e0)

- **Torch**:

![image](https://github.com/user-attachments/assets/b40a8a7e-8206-450d-95d1-9cf708708640)

**Comment by ghost** [comment]
Hey @neel04 

I think with newer jax versions the difference in negligeable:

(5060 Ti)

Jax:

<img width="1363" height="427" alt="Image" src="https://github.com/user-attachments/assets/9a29752f-bc08-42e2-8cf7-37a680ba2b26" />

Torch:

<img width="1320" height="406" alt="Image" src="https://github.com/user-attachments/assets/11d720fc-3346-41ae-ae2b-4b95c4994638" />

**Comment by MoFHeka** [comment]
Any update? May I ask if jax is still persistently optimizing GPU performance? Or is it only focused on TPU?

**Comment by sbodenstein** [comment]
@MoFHeka: note that our production kernels will be here (https://github.com/openxla/tokamax). Definitely still optimizing GPU performance.

**Comment by Rick0827** [comment]
> [@MoFHeka](https://github.com/MoFHeka): note that our production kernels will be here (https://github.com/openxla/tokamax). Definitely still optimizing GPU performance.

I found an awesome impl of flex attention in `tokamax/_src/ops/flex_attention`. Is there any plan to add backward support in the future?

**Comment by sbodenstein** [comment]
@chr1sj0nes

**Comment by MoFHeka** [comment]
> note that our production kernels will be here (https://github.com/openxla/tokamax). Definitely still optimizing GPU performance.

@sbodenstein Any benchmark result available? How does the performance of flash_attention in tokamax compare with that of Nvidia transformer_engine and Flash Attention 3?

**Comment by sbodenstein** [comment]
Not yet. We will have these in time for the hard launch of the library in ~2-3 months.

**Comment by MoFHeka** [comment]
@sbodenstein Hi~! Any progress?

**Comment by janpfeifer** [comment]
Just FYI/out-of-curiosity, here an updated table generated using an RTX 5090 (Blackwell):

|   B |    T |   H |   C |   JAX Naive |   JAX Pallas |   JAX DPA |   PyTorch Naive (Eager) |   PyTorch Naive (Compiled) |   PyTorch SDPA (Eager) |   PyTorch SDPA (Compiled) |
|-----|------|-----|-----|-------------|--------------|-----------|-------------------------|----------------------------|------------------------|---------------------------|
|   8 | 1024 |  16 |  64 |       41.50 |       130.88 |    126.29 |                   29.91 |                     178.52 |                 180.03 |                    182.85 |
|   8 | 1024 |  16 | 128 |       70.80 |       143.76 |    157.00 |                   50.52 |                      79.23 |                 165.63 |                    173.39 |
|   8 | 1024 |  32 |  64 |       39.95 |       146.83 |    160.95 |                   30.32 |                      44.80 |                 209.64 |                    201.26 |
|   8 | 1024 |  32 | 128 |       69.34 |       167.08 |    183.88 |                   52.41 |                      79.33 |                 192.37 |                    191.04 |
|   8 | 2048 |  16 |  64 |       44.59 |       158.54 |    198.17 |                   30.57 |                      46.05 |                 212.93 |                    208.60 |
|   8 | 2048 |  16 | 128 |       80.49 |       180.08 |    204.94 |                   53.44 |                      76.49 |                 199.54 |                    197.76 |
|   8 | 2048 |  32 |  64 |       43.51 |       179.16 |    208.80 |                   30.55 |                      45.66 |                 218.72 |                    214.97 |
|   8 | 2048 |  32 | 128 |       80.90 |       191.33 |    218.16 |                   53.25 |                      76.84 |                 205.01 |                    202.14 |
|  16 | 1024 |  16 |  64 |       40.36 |       152.42 |    161.50 |                   28.45 |                      46.22 |                 201.63 |                    202.89 |
|  16 | 1024 |  16 | 128 |       68.40 |       173.04 |    185.65 |                   52.09 |                      74.42 |                 192.89 |                    188.52 |
|  16 | 1024 |  32 |  64 |       40.30 |       171.19 |    184.36 |                   30.22 |                      45.17 |                 217.90 |                    209.64 |
|  16 | 1024 |  32 | 128 |       69.60 |       182.08 |    200.02 |                   52.09 |                      73.50 |                 195.02 |                    192.75 |
|  16 | 2048 |  16 |  64 |       44.46 |       176.61 |    206.87 |                   30.18 |                      45.54 |                 220.46 |                    215.42 |
|  16 | 2048 |  16 | 128 |       79.97 |       193.15 |    216.80 |                   52.88 |                      74.86 |                 205.85 |                    204.00 |
|  16 | 2048 |  32 |  64 |       44.25 |       187.52 |    216.60 |                   30.42 |                      35.24 |                 216.10 |                    213.29 |
|  16 | 2048 |  32 | 128 |       79.86 |       200.90 |    222.67 |                   52.81 |                      61.73 |                 206.03 |                    204.14 |

- **Versions**: JAX: 0.11.0.dev20260624+4f21d289a; jaxlib: 0.10.2; PyTorch: 2.12.1+cu130; CUDA: 13.0; cuDNN: 9.2.0; Triton: 3.7.1; 
 NVIDIA GPU Driver:  595.71.05;
